### PR TITLE
AP_Baro: Improve HIL and SITL 

### PR DESCRIPTION
**Pull request by patrickelectric**
HIL - Add model for underwater 
SITL - Add BARO_TYPE_WATER functionality

### CODE REVIEW

**Comment by patrickelectric** [comment]
This also solves a problem with barometer initialization in Sub, when checking for BARO_TYPE_WATER.
Sub SITL was running with BARO_TYPE_AIR all this time.

File: libraries/AP_Baro/AP_Baro_HIL.cpp [hunk 1]
```diff
@@ -40,6 +40,45 @@ void AP_Baro::SimpleAtmosphere(
...
+    } else {
+        rho += maxSeaDensity;
+    }
+    rho = rho/seaDensity;
+
+    // From: https://www.grc.nasa.gov/www/k-12/WindTunnel/Activities/fluid_pressure.html
+    // \f$P = \rho (kg) \cdot gravity (m/s2) \cdot depth (m)\f$
+    // \f$P_{atmosphere} = 101.325 kPa\f$
+    // \f$P_{total} = P_{atmosphere} + P_{fluid}\f$
+    float pAtm = 101.325e3;
```
Comment (jaxxzer) [inline libraries/AP_Baro/AP_Baro_HIL.cpp]: Just put 101325 this is misleading with units on first glance. Also, make the units in the code and the comments match.

File: libraries/AP_Baro/AP_Baro_SITL.cpp [hunk 2]
```diff
@@ -103,6 +108,16 @@ void AP_Baro_SITL::_timer()
 
     temperature_adjustment(p, T);
 
+    #else
+
+    float rho, delta, theta;
+    const float p0 = 101325.0f;
+    AP_Baro::SimpleUnderWaterAtmosphere(-sim_alt * 0.001f, rho, delta, theta);
+    float p = p0 * delta;
+    float T = 303.16f * theta - 273.16f;  // Assume 30 degrees at sea level - converted to degrees Kelvin
```
Comment (jaxxzer) [inline libraries/AP_Baro/AP_Baro_SITL.cpp]: are we compensating for temperature affect on water density ?!

**Comment by jaxxzer** [comment]
This looks generally ok and like it will help us with sitl. For this patch, LGTM.

As a followup addressing these special needs for ardusub, I'm wondering if we should consider subclassing Baro for the separation of air/water logic. Thoughts?

File: libraries/AP_Baro/AP_Baro_HIL.cpp [hunk 3]
```diff
@@ -40,6 +40,44 @@ void AP_Baro::SimpleAtmosphere(
...
+    // \f$P_{total} = P_{atmosphere} + P_{fluid}\f$
+    float pAtm = 101325; // Pa
+    float g = 9.8f; // gravity
+    delta = (pAtm + (seaDensity*1e3)*g*(alt*1e3))/pAtm;
+
+    // From: http://residualanalysis.blogspot.com.br/2010/02/temperature-of-ocean-water-at-given.html
+    // \f$T(D)\f$ Temperature underwater at given temperature
+    // \f$S\f$ Surface temperature at the surface
+    // \f$T(D)\approx\frac{S}{1.8 \cdot 10^{-4} \cdot S \cdot T + 1}\f$
+    float seaTempSurface = 15.0f; // Celsius
```
Comment (jaxxzer) [inline libraries/AP_Baro/AP_Baro_HIL.cpp]: I see we are adjusting temperature for depth, that's cool

File: libraries/AP_Baro/AP_Baro_HIL.cpp [hunk 4]
```diff
@@ -40,6 +40,44 @@ void AP_Baro::SimpleAtmosphere(
...
+{
+    // Values and equations based on:
+    // https://en.wikipedia.org/wiki/Standard_sea_level
+    float seaDensity = 1.024f;      // g/cm3
+    float maxSeaDensity = 1.028f;   // g/cm3
+    float pAC = maxSeaDensity - seaDensity; // pycnocline angular coefficient
+
+    // From: https://www.windows2universe.org/earth/Water/density.html
+    rho = seaDensity;
+    if(alt < 1.0f) {
```
Comment (khancyr) [inline libraries/AP_Baro/AP_Baro_HIL.cpp]: miss a space between if and (

**Comment by patrickelectric** [comment]
Style corrections done.

**Comment by jaxxzer** [comment]
Any dissenting opions @ArduPilot/ardupilot-maintainers?

File: libraries/AP_Baro/AP_Baro_SITL.cpp [hunk 5]
```diff
@@ -61,6 +65,7 @@ void AP_Baro_SITL::_timer()
     // add baro glitch
     sim_alt += _sitl->baro_glitch;
 
+#if !APM_BUILD_TYPE(APM_BUILD_ArduSub)
```
[the remark this answers is not in the document]
Comment (jaxxzer) [reply libraries/AP_Baro/AP_Baro_SITL.cpp]: I think the problem with doing it on type is with sitl, the instance is always air? Should we change that bit instead if it's the case?

[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/AP_Baro/AP_Baro_SITL.cpp]: This PR is setting it: https://github.com/ArduPilot/ardupilot/pull/8320/files#diff-52bebeedffc05919f687a1fde40790aeR21

File: libraries/AP_Baro/AP_Baro_HIL.cpp [hunk 6]
```diff
@@ -40,6 +40,44 @@ void AP_Baro::SimpleAtmosphere(
...
+        rho += pAC;
+    }
+    rho = rho/seaDensity;
+
+    // From: https://www.grc.nasa.gov/www/k-12/WindTunnel/Activities/fluid_pressure.html
+    // \f$P = \rho (kg) \cdot gravity (m/s2) \cdot depth (m)\f$
+    // \f$P_{atmosphere} = 101.325 kPa\f$
+    // \f$P_{total} = P_{atmosphere} + P_{fluid}\f$
+    const float pAtm = 101325; // Pa
+    const float g = 9.8f; // gravity
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply libraries/AP_Baro/AP_Baro_HIL.cpp]: Great, ty for pointing this !

[the remark this answers is not in the document]
Comment (patrickelectric) [reply libraries/AP_Baro/AP_Baro_HIL.cpp]: Done.

File: libraries/AP_Baro/AP_Baro_SITL.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (patrickelectric) [reply libraries/AP_Baro/AP_Baro_SITL.cpp]: Done.

[the remark this answers is not in the document]
Comment (patrickelectric) [reply libraries/AP_Baro/AP_Baro_SITL.cpp]: Sorry for not saying anything about this, the idea to use #if and not type check is to allow a faster performance, without a type set function I do not see a type check necessary, we are already doing this in the constructor.

File: libraries/AP_Baro/AP_Baro_SITL.cpp [hunk 7]
```diff
@@ -94,14 +98,22 @@ void AP_Baro_SITL::_timer()
         sim_alt = _buffer[best_index].data;
     }
 
-    float sigma, delta, theta;
     const float p0 = 101325.0f;
```
Comment (jaxxzer) [inline libraries/AP_Baro/AP_Baro_SITL.cpp]: the #ifdef can go above the const float?

**Pull request merged**